File: modules/project/src/main/java/com/falsepattern/zigbrains/project/platform/ZigDirectoryProjectGenerator.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.platform;

import com.falsepattern.zigbrains.project.ide.newproject.ZigProjectConfigurationData;
import com.falsepattern.zigbrains.project.ide.project.ZigDefaultTemplate;
import com.falsepattern.zigbrains.project.ide.util.projectwizard.ZigProjectSettingsStep;
import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettingsService;
import com.falsepattern.zigbrains.zig.Icons;
import com.intellij.facet.ui.ValidationResult;
import com.intellij.ide.util.projectWizard.AbstractNewProjectStep;
import com.intellij.ide.util.projectWizard.CustomStepProjectGenerator;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.NlsContexts;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.wm.impl.welcomeScreen.AbstractActionWithPanel;
import com.intellij.platform.DirectoryProjectGenerator;
import com.intellij.platform.ProjectGeneratorPeer;
import com.intellij.util.ResourceUtil;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Icon;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

public class ZigDirectoryProjectGenerator implements DirectoryProjectGenerator<ZigProjectConfigurationData>,
        CustomStepProjectGenerator<ZigProjectConfigurationData> {

    @Override
    public @NotNull @NlsContexts.Label String getName() {
        return "Zig";
    }

    @Override
    public @Nullable Icon getLogo() {
        return Icons.ZIG;
    }

    @Override
    public @NotNull ProjectGeneratorPeer<ZigProjectConfigurationData> createPeer() {
        return new ZigProjectGeneratorPeer();
    }

    @Override
    public @NotNull ValidationResult validate(@NotNull String baseDirPath) {
        return ValidationResult.OK;
    }

    private static String getResourceString(String path) throws IOException {
        byte[] data = ResourceUtil.getResourceAsBytes(path, ZigDirectoryProjectGenerator.class.getClassLoader());
        if (data == null)
            throw new IOException("Could not find resource " + path + "!");
        return new String(data, StandardCharsets.UTF_8);
    }

    @Override
    public void generateProject(@NotNull Project project, @NotNull VirtualFile baseDir, @NotNull ZigProjectConfigurationData data, @NotNull Module module) {
        val settings = data.settings();

        var svc = ZigProjectSettingsService.getInstance(project);
        val toolchain = settings.toolchain();
        svc.getState().setToolchain(toolchain);

        val template = data.selectedTemplate();

        if (template instanceof ZigDefaultTemplate.ZigInitTemplate) {
            if (toolchain == null) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Tried to generate project with zig init, but zig toolchain is invalid!",
                                                          NotificationType.ERROR));
                return;
            }
            val zig = toolchain.zig();
            val resultOpt = zig.callWithArgs(baseDir.toNioPath(), 10000, "init");
            if (resultOpt.isEmpty()) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "Failed to invoke \"zig init\"!",
                                                          NotificationType.ERROR));
                return;
            }
            val result = resultOpt.get();
            if (result.getExitCode() != 0) {
                Notifications.Bus.notify(new Notification("ZigBrains.Project",
                                                          "\"zig init\" failed with exit code " + result.getExitCode() + "! Check the IDE log files!",
                                                          NotificationType.ERROR));
                System.err.println(result.getStderr());
            }
        } else {
            try {
                val projectName = project.getName();
                WriteAction.run(() -> {
                    for (val fileTemplate : template.fileTemplates().entrySet()) {
                        var fileName = fileTemplate.getKey();
                        VirtualFile parentDir;
                        if (fileName.contains("/")) {
                            val slashIndex = fileName.indexOf('/');
                            parentDir = baseDir.createChildDirectory(this, fileName.substring(0, slashIndex));
                            fileName = fileName.substring(slashIndex + 1);
                        } else {
                            parentDir = baseDir;
                        }
                        val templateDir = fileTemplate.getValue();
                        val resourceData = getResourceString("project-gen/" + templateDir + "/" + fileName + ".template").replace("@@PROJECT_NAME@@", projectName);
                        val targetFile = parentDir.createChildData(this, fileName);
                        VfsUtil.saveText(targetFile, resourceData);
                    }
                });
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    @Override
    public AbstractActionWithPanel createStep(DirectoryProjectGenerator<ZigProjectConfigurationData> projectGenerator, AbstractNewProjectStep.AbstractCallback<ZigProjectConfigurationData> callback) {
        return new ZigProjectSettingsStep(projectGenerator);
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/project/newproject/ZigDirectoryProjectGenerator.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2024 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.newproject

import com.falsepattern.zigbrains.Icons
import com.falsepattern.zigbrains.shared.zigCoroutineScope
import com.intellij.facet.ui.ValidationResult
import com.intellij.ide.util.projectWizard.AbstractNewProjectStep
import com.intellij.ide.util.projectWizard.CustomStepProjectGenerator
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.impl.welcomeScreen.AbstractActionWithPanel
import com.intellij.platform.DirectoryProjectGenerator
import com.intellij.platform.ProjectGeneratorPeer
import com.intellij.platform.ide.progress.withModalProgress
import kotlinx.coroutines.launch
import javax.swing.Icon

class ZigDirectoryProjectGenerator: DirectoryProjectGenerator<ZigProjectConfigurationData>, CustomStepProjectGenerator<ZigProjectConfigurationData> {
    override fun getName(): String {
        return "Zig"
    }

    override fun getLogo(): Icon {
        return Icons.Zig
    }

    override fun createPeer(): ProjectGeneratorPeer<ZigProjectConfigurationData> {
        return ZigProjectGeneratorPeer(true)
    }

    override fun validate(baseDirPath: String): ValidationResult {
        return ValidationResult.OK
    }

    override fun generateProject(project: Project, baseDir: VirtualFile, settings: ZigProjectConfigurationData, module: Module) {
        project.zigCoroutineScope.launch {
            withModalProgress(project, "Generating Project") {
                settings.generateProject(this, project, baseDir, false)
            }
        }
    }

    override fun createStep(
        projectGenerator: DirectoryProjectGenerator<ZigProjectConfigurationData>?,
        callback: AbstractNewProjectStep.AbstractCallback<ZigProjectConfigurationData>?
    ): AbstractActionWithPanel {
        return ZigProjectSettingsStep(projectGenerator!!, callback)
    }
}